src/b64.ml

```diff
@@ -16,77 +19,144 @@
+let uri_safe_alphabet = make_alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_"
+
+let unsafe_set_be_uint16 =
+  if Sys.big_endian
+  then fun t off v -> unsafe_set_uint16 t off v
+  else fun t off v -> unsafe_set_uint16 t off (swap16 v)
+
+exception Out_of_bounds
+
+let get_uint8 t off =
+  if off < 0 || off >= String.length t then raise Out_of_bounds ;
+  unsafe_get_uint8 t off
+
+let padding = int_of_char '='
+
+let encode pad { emap; _ } input =
+  let n = String.length input in
+  let n' = n // 3 * 4 in
+  let res = Bytes.create n' in
+
+  let emap i = Array.unsafe_get emap i in
+
+  let emit b1 b2 b3 i =
+    unsafe_set_be_uint16 res i
+      ((emap (b1 lsr 2 land 0x3f) lsl 8)
+       lor (emap ((b1 lsl 4) lor (b2 lsr 4) land 0x3f))) ;
+    unsafe_set_be_uint16 res (i + 2)
+      ((emap ((b2 lsl 2) lor (b3 lsr 6) land 0x3f) lsl 8)
+       lor (emap (b3 land 0x3f))) in
+
+  let rec enc j i =
+    if i = n then ()
+    else if i = n - 1 then emit (unsafe_get_uint8 input i) 0 0 j
+    else if i = n - 2 then emit (unsafe_get_uint8 input i) (unsafe_get_uint8 input (i + 1)) 0 j
+    else
+    (emit
+       (unsafe_get_uint8 input i)
+       (unsafe_get_uint8 input (i + 1))
+       (unsafe_get_uint8 input (i + 2))
+       j ;
+     enc (j + 4) (i + 3)) in
+
+  let rec unsafe_fix = function
+  | 0 -> ()
+  | i -> unsafe_set_uint8 res (n' - i) padding ; unsafe_fix (i - 1) in
+
+  enc 0 0 ;
+
+  if pad
+  then begin unsafe_fix ((3 - n mod 3) mod 3) ; Bytes.unsafe_to_string res end
+  else Bytes.sub_string res 0 (n / 3 * 4)
+
+let encode ?(pad = true) ?(alphabet = default_alphabet) input = encode pad alphabet input
+
+let error_msgf fmt = Format.ksprintf (fun err -> Error (`Msg err)) fmt
+
+let decode_result { dmap; _ } input =
+  let n = String.length input in
+  let n' = (n / 4) * 3 in
+  let res = Bytes.create n' in
+
+  let emit a b c d i =
+    let x = (a lsl 18) lor (b lsl 12) lor (c lsl 6) lor d in
+    unsafe_set_be_uint16 res i (x lsr 8) ;
+    unsafe_set_uint8 res (i + 2) (x land 0xff) in
+
+  let dmap i =
+    let x = Array.unsafe_get dmap i in
+    if x = none then raise Not_found else x in
+
+  let rec dec j i =
+    if i = n then Some 0
+    else begin
+      let a = dmap (get_uint8 input i) in
+      let b = dmap (get_uint8 input (i + 1)) in
+      let (d, pad) =
+        let x = get_uint8 input (i + 3) in
+        try (dmap x, 0) with Not_found when x = padding -> (0, 1) in
+      let (c, pad) =
+        let x = get_uint8 input (i + 2) in
+        try (dmap x, pad) with Not_found when x = padding && pad = 1 -> (0, 2) in
+
+      emit a b c d j ;
+      if pad = 0
+      then dec (j + 3) (i + 4)
+      else if i + 4 <> n then None
+      else Some pad end in
+
+  match dec 0 0 with
+  | Some pad -> Ok (Bytes.sub_string res 0 (n' - pad))
+  | None -> error_msgf "Wrong padding"
+  | exception Out_of_bounds ->
+      (* appear when [get_uint8] wants to access to an invalid area *)
+      error_msgf "Wrong padding"
```

Comment:
hmm, I recently saw some protocol which use Base64 without padding `=` bytes, so instead of `Zm9vMQ==`, only `Zm9vMQ` is transmitted. I fail to remember where I saw this, but we may want to support such a strange usage

Reply:
This task will be difficult to handle by design when we short-cut by exception the control flow when this case appear. I mean, a `Wrong padding` can appear in different places: when the length of the string is not a multiple of 4.

So any access (any `get_uint8`) can raise this exception. Then, in you task, we need to separate two kinds of error:
* When we want to access to the first and the second part of the chunk (when we get `a` and `b`)
* When we want to access to the third and the four part of the chunk (when we get `d` and `c`)

In the first case, it's an error because we can not emit at least one byte without `a` and `b`.
The second case is the real `Wrong padding` error.

I will try to figure out an other way to handle this kind of input, however, it's not easy to handle it by design.

Reply:
A question, if the input is `Zm9vMQ`, you expect `foo1` as a result or `foo` ?

Reply:
New semantic of `decode` works on what you expect, I think, now.

Reply:
expectation is `foo1`. thanks for your rework, but it has a factor of 2 impact on performance, so in the end not worth doing? or would taking the old code and only apply the following modification be sufficient (and retain performance characteristics, or is it the exception handler which causes the performance difference?):
```OCaml
   let get_uint8 t off =
    try get_uint8 t off with Out_of_bounds -> padding in
```

Reply:
:+1: :heart:

Reply:
since I recently discovered where this `~pad:false` was used, I wanted to drop a reference in here -- https://tools.ietf.org/html/rfc7515#appendix-C (JSON Web Signature), as used and implemented by https://github.com/mmaker/ocaml-letsencrypt (I ported that to the new base64 library in https://github.com/mmaker/ocaml-letsencrypt/pull/17/commits/0ad6eb7a980b689fe547360259cf7e90ea0a0b12). thanks again for this rework!